(SERVER-2896) Update defaults for new cadir

justinstoller opened the pull request:
With Puppet 7 the cadir location will move by default from `$ssldir` to
Puppet Server's confdir. However we want to respect any users that have
already created their cadir in either the old default location or a new
location. We also want to warn if they have it in the old location that
they should migrate the cadir. If users have a custom cadir but it
resides in the ssldir we also want to warn, a cadir within the ssldir
may cause accidental deletion of the cadir.

justinstoller commented:
This needs some tests around the logging in question and the log messages should be unified with whatever is decided on in https://github.com/puppetlabs/puppet/pull/8430

justinstoller commented on lib/puppetserver/ca/config/puppet.rb:
```diff
@@ -46,20 +46,33 @@ def initialize(supplied_config_path = nil)
+            if running_as_root?
+              '/etc/puppetlabs/puppetserver'
+            else
+              "#{ENV['HOME']}/.puppetserver"
```
This is weird, I have it locally, but I'm not sure a local dev environment ever actually uses it. ie, logging goes to stdout the `puppetserver.conf` is in the `./dev` directory of the repo and everything is under `~/.puppetlabs/{etc,opt}`

justinstoller commented on lib/puppetserver/ca/config/puppet.rb:
```diff
@@ -148,6 +160,30 @@ def resolve_settings(overrides = {})
             settings[setting_name] = substitutions[substitution_name] = subbed_value
           end
 
+          configured_cadir = overrides.fetch(:cadir, false)
```
Should probably pull this into its own method to clean up the caller and make testing easier.

justinstoller commented on lib/puppetserver/ca/config/puppet.rb:
```diff
@@ -46,20 +46,33 @@ def initialize(supplied_config_path = nil)
         end
 
-        def load(cli_overrides = {})
+        def load(cli_overrides = {}, logger)
```
This lets us log warnings without cause errors. Alternatively we could populate a `warnings` member that callers can handle like they currently do for `errors`.

Magisus replied:
I think this is fine. To me, when an error happens, we want to do something else to handle it a lot of the time (like exiting 1). But warnings we just want to get logged and then we move on. So it makes sense to me to just do it inline.

justinstoller commented on lib/puppetserver/ca/config/puppet.rb:
```diff
@@ -46,20 +46,33 @@ def initialize(supplied_config_path = nil)
 
-        def user_specific_conf_file
-          user_specific_conf_dir + '/puppet.conf'
+        def user_specific_puppetconf_file
```
How about `user_specific_puppet_config`? I meant to look into that. `confdir` and `config` are the equivalent Puppet settings.

mwaggett replied:
Sounds great!

justinstoller commented on lib/puppetserver/ca/config/puppet.rb:
```diff
@@ -46,20 +46,33 @@ def initialize(supplied_config_path = nil)
+            if running_as_root?
+              '/etc/puppetlabs/puppetserver'
+            else
+              "#{ENV['HOME']}/.puppetserver"
```
It's already a thing, I had it locally. I've been looking into it more and I think it we can safely set it to `~/.puppetlabs` (especially since we're doing this in 7) but we should probably update some _really_ old docs in Puppet Server [here](https://github.com/puppetlabs/puppetserver/blob/6.x/documentation/gems.markdown#installing-and-removing-gems).

justinstoller commented on lib/puppetserver/ca/config/puppet.rb:
```diff
@@ -148,6 +160,30 @@ def resolve_settings(overrides = {})
+          else
+            old_cadir = File.join(ssldir, 'ca')
+            new_cadir = File.join(user_specific_puppetserver_confdir, 'ca')
+            if File.exist?(old_cadir)
```
Good catch!

justinstoller commented on spec/puppetserver/ca/action/clean_spec.rb:
```diff
@@ -12,6 +12,12 @@
 
   subject { Puppetserver::Ca::Action::Clean.new(logger) }
 
+  def without_possible_deprecation(io)
+    io.each_line.reject do |line|
+      line =~ /Found existing CA directory at the old location/
+    end
+  end
+
```
yeah, the tests in the file want to test that certain things they do _don't_ cause warnings/errors, so for those assumptions to still be valid we want to remove potential warnings from the cadir location.

justinstoller replied:
Open to other ideas for names? `without_possible_cadir_deprecation`? `maybe_remove_cadir_deprecation`? We could remove `maybe` or `possible` to be more terse, though if the old cadir exists and warns during these tests depends on the user running the tests, iirc.

mwaggett replied:
ahhh okay, so the tests are saying "this doesn't cause warnings/errors", but they WILL cause the deprecation warning, so we just strip those out because we don't care about them? okay. Can we call this something like `remove_deprecation_warning` or something like that? I think that would make the intention clearer to me.

then my question is, why is the deprecation warning going to stderr? Is that just what warnings do by default?

justinstoller replied:
what did you think of `without_cadir_deprecation`?

And, yeah, warnings and errors go to stderr.

mwaggett replied:
I like using a verb so that it's clear what the intention is. `remove_cadir_deprecation`?

justinstoller commented on spec/puppetserver/ca/action/revoke_spec.rb:
```diff
@@ -69,7 +69,7 @@
       expect(code).to eq(0)
       expect(stdout.string.chomp).to eq('Revoked certificate for foo')
-      expect(stderr.string).to be_empty
+      expect(stderr.each_line.reject{|l| l =~ /Found existing CA directory at the old location/}).to be_empty
```
Yeah, I thought about that, it didn't really fit in any existing utils files and there was only two places for it to be, so I didn't but I'm happy to once we work out a name.

justinstoller commented on spec/spec_helper.rb:
```diff
@@ -4,6 +4,7 @@
 Bundler.require(:default)
 
 require "puppetserver/ca"
+require 'tmpdir'
```
When running some tests in isolation I saw order dependent failures from not having this library already required. This ensures that `tmpdir` is required before any test files are ran.

justinstoller commented:
Okay, think I've addressed the feedback. Let me know what you think. I think this is still waiting on the message to solidify in the equivalent Puppet PR however.

Here's it in action locally:
```
puppetserver-ca-cli (SERVER-2896 $) :: ./exe/puppetserver-ca list --all
Found existing CA directory at the old location of '/Users/justin/.puppetlabs/etc/puppet/ssl/ca'. This is deprecated and will be moved to '/Users/justin/.puppetlabs/etc/puppetserver/ca' in a subsequent release.
Signed Certificates:
    localhost       (SHA256)  9B:07:8B:52:B5:DB:E5:FB:C9:5E:E4:37:53:76:E6:1E:EA:01:DF:76:D6:77:07:76:41:1B:4C:E8:F0:34:A3:F9   alt names: ["DNS:puppet", "DNS:localhost"]      authorization extensions: [pp_cli_auth: true]
```

justinstoller commented:
Updated with the new error message. I think this is ready for another review.

Magisus reviewed (approved):
This looks good me. 

Do we think it's worth a quick manual test? If so, it'd be good to try it in both an existing CA scenario and to create a new CA at the new location.

Alternatively, I'm fine waiting on that and just using beaker tests to test this end-to-end later once all the bits are in.

mwaggett commented on lib/puppetserver/ca/config/puppet.rb:
```diff
@@ -34,32 +34,46 @@ def self.parse(config_path)
           @errors = []
         end
 
+        def running_as_root?
+          @as_root ||= Puppetserver::Ca::Utils::Config.running_as_root?
+        end
```
Why do we need this instead of calling `Puppetserver::Ca::Utils::Config.running_as_root?` directly? Is it just an expensive thing to check?

justinstoller replied:
I don't remember if it's expensive, though previously we were caching the end result of the call so I preserved that behavior.

mwaggett replied:
We were? It looks like we were previously calling `Puppetserver::Ca::Utils::Config.running_as_root?` just once in this file and not caching that value anywhere?

Magisus replied:
Looks like we were caching the answer we got as a result of the conditional in which this was called. So we wouldn't have called this more than once.

mwaggett commented on lib/puppetserver/ca/config/puppet.rb:
```diff
@@ -210,6 +229,33 @@ def parse_text(text)
+        def find_cadir(configured_cadir, ssldir, logger)
+          warning = 'The cadir is currently configured to be inside the ' +
+            '%{ssldir} directory. This config setting and the directory ' +
+            'location will not used in a future version of puppet. Please ' +
```
```suggestion
            'location will not be used in a future version of puppet. Please ' +
```

mwaggett commented on lib/puppetserver/ca/config/puppet.rb:
```diff
@@ -148,6 +161,12 @@ def resolve_settings(overrides = {})
           end
 
+          cadir = find_cadir(overrides.fetch(:cadir, false),
+                             settings.fetch(:ssldir, '/etc/puppetlabs/puppet/ssl'),
```
Does it make sense to use `$confdir/ssl` here instead of hard-coding `/etc/puppetlabs/puppet/ssl`?
I'm assuming that param is what's used if the setting doesn't exist and I'm not really sure how it wouldn't exist since we set it above?

justinstoller replied:
It doesn't make sense, but also in thinking through that reply I realized that this default isn't needed either.

mwaggett reviewed (commented):
Can we test that the `cadir` setting is set to the appropriate values in each situation?

justinstoller commented:
Addressed feedback
